src/Data/TaskMap.hs

```diff
@@ -88,14 +96,14 @@ memo
     -> (k -> IO v)
         -- ^ an action that is used to produce the value if the key isn't in the map.
     -> IO v
-memo tm@(TaskMap var) k task = do
-    -- TODO: should we insert another lookup here? It depends on how optimistic
-    -- we are that the lookup is successful.
-
-    a <- modifyMVarMasked var $ \m -> case HM.lookup k m of
-        Nothing -> do
-            !a <- asyncWithUnmask $ \umask -> umask (task k) `finally` delete tm k
-            m' <- evaluate $ HM.insert k a m
-            return (m', a)
-        (Just !a) -> return (m, a)
-    wait a
+memo tm@(TaskMap var) k task = bracket query cancel wait
+  where
+    query = do
+        -- NOTE: should we insert another lookup here? It depends on how optimistic
+        -- we are that the lookup is successful.
+        modifyMVarMasked var $ \m -> case HM.lookup k m of
+            Nothing -> do
+                !a <- asyncWithUnmask $ \umask -> umask (task k) `finally` delete tm k
+                !m' <- evaluate $ HM.insert k a m
```

Comment:
```suggestion
                m' <- evaluate $ HM.insert k a m
```

Let's get this PR merged -- it fixes a bug and should not stall and bitrot just because of Haskell code style questions.